### PR TITLE
Implement transparent new-types in `ssz_derive`

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

Related to #3625

## Proposed Changes

Allows `transparent` encoding for new-type structs, just like `serde` does.

```rust
/// The encoded/decoded bytes will be identical to a `Vec<u8>`
#[derive(PartialEq, Debug, Encode, Decode)]
#[ssz(struct_behaviour = "transparent")]
struct TransparentStruct {
    inner: Vec<u8>,
}
```

## Additional Info

I added some top-level docs to `ssz_derive` to make sure we're keeping track of all the attributes somewhere.

I've also moved the tests for the derive macro from `eth2_ssz` to `eth2_ssz_derive`. I did this primarily because `cargo-udeps` can't detect when a dev-dependency is only used in doc-tests and raises a false positive. Moving the tests caused `eth2_ssz` to be used in tests and therefore udeps was happy again (there's an ignore method for udeps but it was too blunt IMO). It also kinda makes sense to have those tests in the derive crate rather than where they were.

### CODE REVIEW

**Comment by paulhauner** [comment]
FYI @realbigsean here's that SSZ impl we discussed last week :relaxed: Does it have all the features you need?

IIRC, you need something like this (which is supported):

```rust
/// The encoded/decoded bytes will be identical to a `Vec<u8>`
#[derive(PartialEq, Debug, Encode, Decode)]
#[ssz(struct_behaviour = "transparent")]
struct TransparentStruct {
    serialized_thing:: Vec<u8>,
    #[ssz(skip_serializing, skip_deserializing)]
    local_thing: u8,
}
```

File: consensus/ssz_derive/src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,123 @@
 #![recursion_limit = "256"]
 //! Provides procedural derive macros for the `Encode` and `Decode` traits of the `eth2_ssz` crate.
 //!
-//! Supports field attributes, see each derive macro for more information.
+//! ## Attributes
+//!
+//! The following struct\enum attributes are available:
```
Comment (realbigsean) [inline consensus/ssz_derive/src/lib.rs]:
```suggestion
//! The following struct/enum attributes are available:
```

File: consensus/ssz_derive/src/lib.rs [hunk 2]
```diff
@@ -545,6 +761,81 @@ fn ssz_decode_derive_struct(item: &DeriveInput, struct_data: &DataStruct) -> Tok
     output.into()
 }
 
+/// Implements `ssz::Decode` "transparently" for a `struct` with exactly one non-skipped field.
+///
+/// The bytes will be decoded as if they are the inner field, without the outmost struct. The
```
Comment (realbigsean) [inline consensus/ssz_derive/src/lib.rs]:
```suggestion
/// The bytes will be decoded as if they are the inner field, without the outermost struct. The
```

File: consensus/ssz_derive/src/lib.rs [hunk 3]
```diff
@@ -545,6 +761,81 @@ fn ssz_decode_derive_struct(item: &DeriveInput, struct_data: &DataStruct) -> Tok
...
+        .count();
+
+    if num_fields != 1 {
+        panic!(
+            "A \"transparent\" struct must have exactly one non-skipped field ({} fields found)",
+            num_fields
+        );
+    }
+
+    let mut field_names = vec![];
```
Comment (realbigsean) [inline consensus/ssz_derive/src/lib.rs]: I don't think these `field_names` are used

**Review by realbigsean — COMMENTED** [review]
Looks awesome! The new docs are great. I took a stab at also supporting unnamed fields here:
https://github.com/realbigsean/lighthouse/commit/697e190aa6fac188a231a847f53dbd3dc1aa1496

I included the suggested changes below in that commit ^

File: consensus/ssz_derive/src/lib.rs (hunk 3 above)
Comment (paulhauner) [reply consensus/ssz_derive/src/lib.rs]: Nice catch!

**Comment by paulhauner** [comment]
> I took a stab at also supporting unnamed fields 

That was great, I included it in the PR! I made a change in 27a5a36bc to support the scenario where there are skipped tuple fields *before* the non-skipped field.

This is ready for review again :relaxed:

**Comment by paulhauner** [comment]
Closed in favor of #3861

**Pull request closed without merging**